File: src/python/ted/TEDataFile.py
import copy
from pathlib import Path

import pandas as pd

from src.python.path import pathOfTEDFile
from src.python.config.config import flowTypes, techs, mapColnamesDtypes
from src.python.ted.inconsistencies import checkRowConsistency


class TEDataFile:
    # initialise
    def __init__(self, tid: str, path: None|Path = None):
        self._tid: str = tid
        self._tspecs: dict = copy.deepcopy(techs[tid])
        self._path: Path = path or pathOfTEDFile(self._tid)
        self._df: None | pd.DataFrame = None
        self._inconsistencies: dict = {}


    # load TEDataFile (only if it has not been read yet)
    def load(self):
        if self._df is None:
            self.read()


    # read TEDataFile from CSV file
    def read(self):
        # read CSV file
        self._df = pd.read_csv(
            self._path,
            dtype=mapColnamesDtypes,
            sep=',',
            quotechar='"',
            encoding='utf-8',
        )

        # adjust row index to start at 1 instead of 0
        self._df.index += 1

        # make sure the file contains no unknown columns
        dataFormatColIDs = list(mapColnamesDtypes.keys())
        for colID in self._df.columns:
            if colID not in dataFormatColIDs:
                raise Exception(f"Unknown column '{colID}' in file \"{self._path}\".")

        # insert missing columns and reorder via reindexing
        self._df = self._df.reindex(columns=dataFormatColIDs)


    # write TEDataFile to CSV file
    def write(self):
        self._df.to_csv(
            self._path,
            index=False,
            sep=',',
            quotechar='"',
            encoding='utf-8',
            na_rep='',
        )


    # get data
    @property
    def data(self) -> pd.DataFrame:
        return self._df


    # get data
    def getInconsistencies(self) -> dict:
        return self._inconsistencies


    # check that TEDataFile is consistent
    def check(self, re: bool = True):
        self._inconsistencies = {}

        # check row consistency for each row individually
        for rowID in self._df.index:
            self.checkRow(rowID, re=re)


    # check that row in TEDataFile is consistent
    def checkRow(self, rowID: int, re: bool = True):
        row = self._df.loc[rowID]
        self._inconsistencies[rowID] = checkRowConsistency(self._tid, row, re=re, rowID=rowID, filePath=self._path)

File: R/ted/TEDataFile.R
source("R/ted/TEBase.R")

TEDataFile = R6Class("TEDataFile",
  inherit = TEBase,
  private = list(
      tid = NULL,
      path = NULL,
      df = NULL
  ),
  public = list(
    initialize = function(tid, path=NULL) {
      # initialise TEBase fields
      super$initialize(tid)
      # initialise object fields
      private$tid <- tid
      if (is.null(path)){
          private$path <- pathOfTEDFile(tid)
      } else {
          private$path <- path
      }
      private$df <- data.frame(NULL)
    },
    load = function() {
      if (identical(private$df,data.frame(NULL))) {
        self$read()
        return(self)
      }
    },
    read = function() {
      # read CSV file
      print(sprintf("Reading file \"%s\"...", private$path))
      private$df <- read.csv(
          private$path,
          colClasses=mapColnamesDtypes,
          header=TRUE,
          sep=",",
          quote="\"",
          encoding="UTF-8",
          na.strings=""
      )
      #private$df <- fread(
      #    private$path,
      #    colClasses=mapColnamesDtypes,
      #    #header=TRUE,
      #    sep=",",
      #    quote='"',
      #    encoding="UTF-8",
      #    na.strings="",
      #    data.table=FALSE
      #)
      # adjust row index to start at 1 instead of 0
      rownames(private$df) <- seq_len(nrow(private$df))

      # make sure the file contains no unknown columns
      dataFormatColIDs <- names(private$dataFormat)
      for (colID in names(private$df)) {
          if (!(colID %in% dataFormatColIDs)) {
              stop(sprintf("Unknown column '%s' in file \"%s\".", colID, private$path))
          }
      }
      
      # insert missing columns and reorder via reindexing and update dtypes
      dfNew <- private$df[, dataFormatColIDs]
      cols <- names(self$getDtypeMapping())
      dtypes <- self$getDtypeMapping()
      for (i in seq_along(cols)) {
          col <- cols[[i]]
          dtype <- dtypes[[i]]
          if (col %in% names(private$df)) {
              next
          }
          dfNew[[col]] <- dfNew[[col]] %>% type.convert(as.is=TRUE)
          dfNew[[col]] <- NA
      }
      private$df <- dfNew
    },
    write = function() {
      write.csv(
          private$df,
          file=private$path,
          row.names=FALSE,
          sep=",",
          qmethod="escape",
          fileEncoding="UTF-8",
          na=""
      )
    },
    data = function() {
      private$df
    },
    getInconsistencies = function() {
      private$inconsistencies
    },
    check = function(re = TRUE) {
      private$inconsistencies <- list()

      # check row consistency for each row individually
      for (rowID in rownames(private$df)){
          checkRow(.Object, rowID, re)
      }
    },
    checkRow = function(rowID, re = TRUE) {
      row <- private$df[rowID,]
      # consistency checks are not implemented in R for now
      # private$inconsistencies[[rowID]] <- checkRowConsistency(private$tid, row, re, rowID, private$path) # nolint # nolint: line_length_linter.
    }
  )
)